src/common/transformations/src/transformations/common_optimizations/convert_nms_gather_path_to_unsigned.cpp

```diff
@@ -60,14 +62,62 @@ class PropagateNMSPath : public pass::MatcherPass {
                                                ov::op::v1::VariadicSplit,
                                                op::util::GatherBase,
                                                ov::op::v0::Concat,
-                                               ov::op::v0::Convert>();
+                                               ov::op::v0::Convert,
+                                               ov::op::v8::If>();
         matcher_pass_callback callback = [=](pattern::Matcher& m) {
+            auto handle_if_params =
+                [](std::shared_ptr<ov::op::v8::If> if_node, std::shared_ptr<ov::Model> body, int body_index) {
+                    const auto& params = body->get_parameters();
+                    for (auto input_desc : if_node->get_input_descriptions(body_index)) {
+                        auto param = params[input_desc->m_body_parameter_index];
+                        auto input_node = if_node->input(input_desc->m_input_index).get_source_output().get_node();
+
+                        if (ov::has_nms_selected_indices(input_node)) {
+                            ov::set_nms_selected_indices(param.get());
+                        }
+                    }
+                };
+            auto handle_if_results =
+                [](std::shared_ptr<ov::op::v8::If> if_node, std::shared_ptr<ov::Model> body, int body_index) {
+                    const auto& results = body->get_results();
+                    for (auto output_desc : if_node->get_output_descriptions(body_index)) {
+                        auto result = results[output_desc->m_body_value_index];
+                        const auto& result_inputs = result->input_values();
+                        auto output_node = if_node->output(output_desc->m_output_index).get_node();
+
+                        if (any_of(result_inputs.begin(), result_inputs.end(), [](const Output<Node>& output) {
+                                return ov::has_nms_selected_indices(output.get_node());
+                            })) {
+                            ov::set_nms_selected_indices(output_node);
+                        }
+                    }
+                };
+
             auto node = m.get_match_root();
-            const auto& inputs = node->input_values();
-            if (any_of(inputs.begin(), inputs.end(), [](const Output<Node>& output) {
-                    return ov::has_nms_selected_indices(output.get_node());
-                })) {
-                ov::set_nms_selected_indices(node.get());
+
+            if (ov::is_type<ov::op::v8::If>(node)) {
```

Comment:
I believe it's better to cast it to "op::util::MultiSubgraphOp", this is the base class for Loop, TI and IF
after that we will be able to extend and simplify the code 
```
const auto& models = multi_subgraph_op->get_functions();
for (size_t i = 0; i < models.size(); ++i) {
    handle_params(multi_subgraph_op, models[i], i);
    ov::pass::Manager manager;
    manager.register_pass<ov::pass::PropagateNMSPath>();
    manager.run_passes(then_body);
    handle_results(multi_subgraph_op, models[i], i);
}
```
But it's better to leave only "If" operation in the list, not TI and Loop, we don't have real cases for that.